docs/developer/go-style-guide.md

```diff
@@ -0,0 +1,457 @@
+Use _templates_ to substitute for concrete types, use _interfaces_ to substitute for abstract behaviors.
+
+### Do not use booleans as function parameters
+
+A boolean can only express an on/off condition and will require a breaking change or new arguments in the future.
+Instead, pack functionality; for example, using integers instead of bools (maps and structs are also acceptable).
+
+DO NOT:
+
+```go
+func(bool userOn, bool groupOn, bool globalOn)
+```
+
+DO:
+
+```go
+func(uint permissions) // permissions := USER | GROUP | GLOBAL;  if permissions & USER then USER is set
+```
+
+### Use dependency injection to separate concerns
+
+Use dependency injection to separate high-level policy from low-level detail. Dependency injection promotes modular,
+testable, and maintainable code by reducing coupling and increasing flexibility.
+
+Creating dependencies couples ownership and lifetime while making tests difficult or impossible. Constructing
+dependencies adds side-effects which complicates testing.
+
+### Required arguments should be provided via parameters and optional arguments provided functionally or with structs
+
+DO NOT:
+
+```go 
+func(int required, int optional) {
+  if optional {...} 
+}
+```
+
+DO:
+
+```go
+type Option func (o *Object)
+
+func Optional(string optional) Option {
+  return func (o *Object) {
+    o.optional = optional
+  } 
+}
+
+func (int required, ...Options) {
+  for o := range Options {
+    o(self)
+  } 
+}
+```
+
+## Error Handling
+
+### Do not filter context when returning errors
+
+Preserve error context by wrapping errors as the stack unwinds. Utilize native error wrapping with `fmt.Errorf` and
+the `%w` verb to wrap errors. Wrapped errors offer a transparent view to end users. For a practical example, refer to
+this runnable code snippet: [Go Playground Example](https://go.dev/play/p/f9EaJDB5JUO). When required, you can identify
+inner wrapped errors using native APIs such as `As`, `Is`, and `Unwrap`.
+See [Working with errors](https://go.dev/blog/go1.13-errors) for more information on error wrapping.
+
+### Only handle errors once
+
+DO NOT log an error, then subsequently return that error. This creates the potential for multiple error reports and
+conflicting information to users.
+
+DO NOT:
+
+```go
+func badAtStuff(noData string) error {
+  if len(noData) == 0 {
+    fmt.Printf("Received no data")
+  }
+
+  return errors.New("received no data")
+}
+```
+
+DO
+
+```go
+func badAtStuff(noData string) error {
+  if len(noData) == 0 {
+    return errors.New("received no data")
+  }
+  ...
+}
+```
+
+A leader of the golang community once said:
+> Lastly, I want to mention that you should only handle errors once. Handling an error means inspecting the
+> error value, and making a decision. If you make less than one decision, you’re ignoring the error...But making
+> more than one decision in response to a single error is also problematic. - Dave Cheney
+
+### Libraries should return errors for callers to handle
```

Comment:
shall we mention when to use panics? 
currently, we also pass en error to panic, like panic(fmt.Error(...))